diff --git a/src/wifibroadcast.cpp b/src/wifibroadcast.cpp
index e5d44b3..9408dc3 100644
--- a/src/wifibroadcast.cpp
+++ b/src/wifibroadcast.cpp
@@ -136,6 +136,20 @@ int open_unix_socket_for_rx(const char *socket_path, int rcv_buf_size, int socke
 
     memset(&saddr, '\0', sizeof(saddr));
     saddr.sun_family = AF_UNIX;
+#ifdef __APPLE__
+    if (strlen(socket_path) >= sizeof(saddr.sun_path))
+    {
+        close(fd);
+        throw runtime_error("Unix socket path is too long");
+    }
+    strcpy(saddr.sun_path, socket_path);
+    saddr.sun_len = sizeof(saddr);
+    if (::bind(fd, (struct sockaddr *) &saddr, sizeof(saddr)) < 0)
+    {
+        close(fd);
+        throw runtime_error(string_format("Unable to bind to %s : %s", socket_path, strerror(errno)));
+    }
+#else
     strncpy(saddr.sun_path + 1, socket_path, sizeof(saddr.sun_path) - 2);
 
     if (::bind(fd, (struct sockaddr *) &saddr, sizeof(sa_family_t) + strlen(saddr.sun_path + 1) + 1) < 0)
@@ -143,6 +157,7 @@ int open_unix_socket_for_rx(const char *socket_path, int rcv_buf_size, int socke
         close(fd);
         throw runtime_error(string_format("Unable to bind to @%s : %s", saddr.sun_path + 1, strerror(errno)));
     }
+#endif
     return fd;
 }
 
